### ISSUE TITLE
Notify when average ping for last X minutes is higher than threshold

### ISSUE THREAD

**Issue by DmytroSokhach** [louislam/uptime-kuma#5211]
### 📑 I have found these related issues/pull requests

https://github.com/louislam/uptime-kuma/issues/1724
https://github.com/louislam/uptime-kuma/issues/4234

### 🏷️ Feature Request Type

New notification-provider, Other

### 🔖 Feature description

My network provider is quite unstable, thanks to DSL (yes, still used in 2024 🙃)
Happens in about 1-2 weeks, and since home network is sometimes critical for communication I need to get notified before I notice problems while working.

**Example use case:**
ping to google.com jumps from 40-60ms to 120-200ms or even 300-500ms, which are the signs that something is wrong on ISP router side.

### ✔️ Solution

**Proposal:**
Have additional category "Latency alert" (or some better naming, doesn't really matter), which would detect and notify about latency issues.

### ❓ Alternatives

The only option is to "suspect" and then "check", but it is reactive behaviour which I can do by running mtr (or WinMTR) to check path ping.
I would really love to be notified proactively about problem.

**Comment by pmcnano** [comment]
I literally came just searching for this 😄

**Comment by Feriman22** [comment]
I'd request the same. I'd suggest the name of this would be "Ping latency check" or just "Ping latency" next to the Ping:

![Image](https://github.com/user-attachments/assets/f523dfc9-01df-4446-8cee-ff9e9a0a39be)

**Comment by tglucas** [comment]
> I'd ask for the same thing. I'd suggest calling it "ping latency check" or just "ping latency" next to the ping:
> 
> ![Image](https://github.com/user-attachments/assets/f523dfc9-01df-4446-8cee-ff9e9a0a39be)

Perhaps instead, additional parameter fields on the existing Ping monitor configuration? It would be handy to define some kind of timeout or deadline generally.

**Comment by Feriman22** [comment]
> > I'd ask for the same thing. I'd suggest calling it "ping latency check" or just "ping latency" next to the ping:
> > ![Image](https://github.com/user-attachments/assets/f523dfc9-01df-4446-8cee-ff9e9a0a39be)
> 
> Perhaps instead, additional parameter fields on the existing Ping monitor configuration? It would be handy to define some kind of timeout or deadline generally.

It's also a good solution.

**Comment by DmytroSokhach** [comment]
I agree that actual Ping monitor is already good enough AND just needs additional notification trigger.
"**Maximum Latency Notification**" (or better more clear name)

![Image](https://github.com/user-attachments/assets/e51826b3-da1e-40ff-bc8b-5bfb58f598c9)

**Comment by pcgeek86** [comment]
The HTTP(s) monitor type already has a timeout feature on it. 

It would be nice if there were an optional timeout configurable on the ICMP (Ping) monitor as well.

My use case(s) are: 

1. I have a relatively simple network configuration, but sometimes I have an access point that starts behaving "oddly." Any devices that are connected through that access point experience high ping times. I would like to be proactively notified when this admittedly rare behavior occurs. It would simplify network troubleshooting, so I don't start suspecting other devices, or my Starlink ISP, which are not at fault.

1. I use Starlink as an ISP. The vast majority of the time, I get excellent performance, including bandwidth and latency. However, there are pretty rare occasions where I might experience an outage. It would be great to have an ICMP monitor set up that can detect latency spikes that last for more than .... let's say ~30 seconds.

**Comment by Whatsonyourmind** [comment]
The core challenge here is distinguishing genuine latency degradation from normal jitter — a static threshold (e.g., "alert if ping > 200ms") will either miss slow ramp-ups or fire too often on noisy DSL lines.

A better approach is **z-score anomaly detection**: compute a rolling baseline of your recent pings, then flag when the current window deviates by more than N standard deviations. This adapts to your normal range automatically.

Here's the concept with example data:

```
pings = [42, 55, 48, 51, 47, 53, 49, 180, 220, 310, 45, 52]

mean = 55.2
std  = 80.1

z_scores:
  42  → -0.16  (normal)
  180 →  1.56  (borderline)
  220 →  2.06  (ANOMALY at threshold=2.0)
  310 →  3.18  (ANOMALY)
```

The 180/220/310 spike gets flagged while normal 40-55ms jitter passes through. You can tune the threshold — 2.0 catches ~95% of outliers, 2.5 is more conservative.

For implementation: collect the last N pings (say 30 minutes worth), compute the rolling z-score, and trigger a notification if any point exceeds the threshold. This handles your exact use case — detecting ISP degradation *before* full outage.

**Alternative methods** depending on your data characteristics:
- **IQR (interquartile range)**: More robust to existing outliers in your baseline. Flag anything > Q3 + 1.5×IQR.
- **Modified z-score** (using median absolute deviation): Better when your ping distribution is skewed, which DSL lines often are.

Would a rolling-window z-score approach cover your case, or do you also need to handle gradual drift (e.g., ping slowly climbing from 50ms to 150ms over hours)? For drift you'd want a change-point detection algorithm instead.

**Comment by CommanderStorm** [comment]
thanks for the advertisement @Whatsonyourmind (might be relevant, but still we should move towards reviewing and merging the open PR instead..)

**Comment by Whatsonyourmind** [comment]
@CommanderStorm Fair enough -- you're right, the existing PR is the priority. Which PR is open for this? Happy to review or test it if that would help move it forward.

**Comment by CommanderStorm** [comment]
-  https://github.com/louislam/uptime-kuma/pull/4005

**Comment by Whatsonyourmind** [comment]
@CommanderStorm Thanks for the link. I'll take a look at PR #4005 and test if I can.
